### DISCUSSION TITLE
Configurazione per accesso esterno con caddy

### DISCUSSION THREAD

**Discussion by basebuzz** [indomus/forum#3283] [HUB: Home Assistant]
Ciao, ho seguito passo passo questa guida https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reverse-proxy/
accedo al mi HA esternamente via http ma non via https, uso HA e il resto tutto su raspberry. nei log di caddy sembra che sia tutto apposto, non riesco però a venirne a capo.  ho seguito e completato anche la guida https://indomus.it/guide/aggiornare-il-proprio-record-su-duckdns-tramite-automazione-home-assistant/

Avrei bisogno di un aiuto, non so dove sto sbagliando. grazie

**Comment by snakuzzo** [comment]
Perché ti sei perso qualche pezzo.
Se raggiungi HA via http vuol dire che di sicuro hai sbagliato qualcosa a livello di port forwarding sul router.

Se ad esempio hai `miodominio.com` e vuoi che risponda sulla porta `9000` devi fare queste cose:

1. Se non hai il plugin su Caddy per la DNS challenge, devi forwardare la porta 443 verso la 443 o la 80 verso la 80 di caddy per il funzionamento delle challenge http-01 e/o tls-alpn per il rilascio dei certificati

2. Devi forwardare la `9000` verso la `9000` di Caddy, sarà poi Caddy a mandare il traffico verso HA sulla porta 8123 (in base alla configurazione nel ile `Caddyfile`) 

3. Eliminare la regola di Port forwarding che hai creato (perché l'hai fatto) che redirige il traffico verso la `8123` di HA

Ovviamente, ripeto, questo ammesso che tu abbia intenzione di raggiungere HA sulla 9000, in caso contrario adegua in base alle porte che ti servono

**Reply by basebuzz** [reply]
Ciao snakuzzo, grazie per il tuo aiuto.
allora ho fatto come hai detto, partendo dal fatto che voglio che funzioni indipendentemente dalla porta utilizzata, facciamo che debba rispondere sulla 9000. quindi https://xxxxxx.duckdns.org:9000.
ho tolto il redirect sulla 8123 nel router e forwardato la 443 e la 9000 verso caddy (Stessa macchina di HA).
ora se mi scollego dal router, con il telefono e browser se raggiungo https://xxxxxx.duckdns.org:9000 mi ritorna errore 400 bad request. se provo via http://xxxxx.duckdns.org mi da il messaggio pagina bianca e "saluti dal crew di indomus.it".

In realtà ho un paio di dubbi che dalla guida non sono riuscito a capire:
- riguardo file di configurazione di caddy da mettere in /addon_configs/c80c7555_caddy-2/ , la directory /c80c7555_caddy-2/ non esisteva, l'ho creata e ci ho messo dentro il file con le impostazioni come da guida. questo mi puzza un po di cosa strana o sbagliata.
- l'addon DuckDNS in HA va installato? sulla guida non c'è scritto, ho completato i passaggi per l'automazione aggiornamento record di duckdns ma l'addon non l'ho installato. è corretto?
- il mio intento è di raggiungere HA  sia da casa che da fuori ma solo con l'APP HA senza ogni volta intervenire sulla configurazione dell'app o fare altro di "tecnico" perchè devo farla usare anche a mia moglie.
- altra e ultima cosa, in HA settings->sistema->rete, devo impostare qualcosa in particolare? 
grazie 1000 per l'aiuto

**Reply by snakuzzo** [reply]
Quindi stai usando addon su ha os.
Allora...

Il 400 deriva dal fatto che non hai impostato i trusted proxies sotto il blocco http del configutation.yaml. Vedi bene che c'è scritto, anche sulla documentazione dell'addon.

Quanto ai tuoi dubbi...

- la dir dell'addon non ricordo se veniva creata automaticamente oppure no
- se aggiorni già con il package il record DNS, non hai bisogno dell'addon duckdns
- così come stai facendo raggiungi l'obiettivo. Sull'app companion sullo smartphone inserirai URL esterna `https://mio.dominio.com:9000` e URL interna `http://iplocale:8123` (attenzione alla differenza di protocollo http e https).
- sulla configurazione di sistema non devi fare nulla

**Reply by basebuzz** [reply]
Allora nel configurations.yaml ho messo:

```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 127.0.0.1
```

è corretto?

**Reply by snakuzzo** [reply]
Innanzi tutto, quando posti il codice devi usare il markdown... è scritto chiaramente anche sulla root di discussions.
Poi, tornando alla tua domanda, dipende dall'IP con cui Caddy si presenta su Core e questo puoi vederlo proprio nei log di core ogni volta che provi la connettività passando da caddy. Vedrai che ti darà un warning indicando l'IP sorgente.

Io comunque ti direi di impostare così...

```yaml
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 127.0.0.1 # la loopback ipv4
    - 192.168.1.0/24 # da adeguare con la subnet della tua LAN
    - ::1 # la loopback ipv6
    - 172.30.33.0/24 # la subnet docker di tutti gli addon
 ```

**Reply by basebuzz** [reply]
chiedo perdono per non aver usato il markdown. aggiornato il post.
@snakuzzo che dire: grazie, funziona tutto!!

**Reply by basebuzz** [reply]
Ciao @snakuzzo, dopo molti mesi di funzionamento, non riesco più a raggiungere HA dall'esterno, il messaggio di errore dice che la data del certificato di HA non è valida. cosa può essere successo ? grazie

**Reply by snakuzzo** [reply]
è scaduto il certificato e non è stato rinnovato. Aggiorna/riavvia Caddy e controlla i log (di Caddy) per vedere cosa succede

**Reply by basebuzz** [reply]
questa l'ultima riga del log di caddy:
authorization failed: HTTP 400 urn:ietf:params:acme:error:connection - xx.xx.xxx.xxx: Connection refused (ca=[https://acme-v02.api.letsencrypt.org/directory)","attempt":1,"retrying_in":60,"elapsed":4.986928429,"max_duration":2592000}](https://acme-v02.api.letsencrypt.org/directory)%22,%22attempt%22:1,%22retrying_in%22:60,%22elapsed%22:4.986928429,%22max_duration%22:2592000%7D)

**Reply by basebuzz** [reply]
@snakuzzo riesci a darmi una mano? Probabilmente l'indirizzo ip esterno è cambiato ma lo script messo in atto non funziona. Presumo, poi mi posso sbagliare

**Reply by snakuzzo** [reply]
Devi assicurarti che l'indirizzo IP pubblico che ti è stato assegnato sia lo stesso che viene restituito quando viene risolto il nome dai DNS pubblici

**Reply by snakuzzo** [reply]
Per intenderci...

```
curl ipinfo.io/ip
```

ti deve restituire lo stesso IP di una nslookup

```
nslookup <tuodominio>
```

**Reply by basebuzz** [reply]
> Per intenderci...
> 
> ```
> curl ipinfo.io/ip
> ```
> 
> ti deve restituire lo stesso IP di una nslookup
> 
> ```
> nslookup <tuodominio>
> ```

entrambe restituiscono lo stesso ip che è effettivamente il mio ip pubblico, non capisco il problema.

**Reply by snakuzzo** [reply]
Ma prima di andare avanti... hai letto il mio primo messaggio ?
Ti avevo scritto delle cose e tu hai continuato a pensare a Caddy.
Tu hai detto che raggiungi dall'esterno HA in HTTP e NON in HTTPS. Questa cosa già di per se non va bene. Se non capisci bene il giro è inutile proseguire.

HA deve essere esposto in HTTP solo su rete interna, non devi aprire alcuna porta verso la 8123 di HA.

Porta esterna scelta da te (es. 9000) -> Porta di Caddy (es. 9000) -> deve proxare verso IP di HA su porta di HA (es. 192.168.1.10:8123)

Un esempio del Caddyfile...

```
(https_header_core) {
	header {
		Strict-Transport-Security "max-age=31536000; includeSubdomains"
		X-XSS-Protection "1; mode=block"
		X-Content-Type-Options "nosniff"
		X-Frame-Options "SAMEORIGIN"
		Referrer-Policy "same-origin"
	}
} 

https://tuodominio:9000 {
	import https_header_core
	reverse_proxy http://192.168.1.10:8123
        @ws {
            header Connection *Upgrade*
            header Upgrade websocket
        }
}
```

e come dovrebbe essere impostato il porforwarding sul router...

- porta esterna 9000 TCP -> 192.168.1.10 porta 9000 TCP  (che è la porta di ascolto del servizio che dedichi ad HA)

poi...puoi impostare uno di questi due o anche tutti e due...servono alle challenge per i certificati:

- porta esterna 443 TCP -> 192.168.1.10 porta 443 TCP (che è la porta di Caddy usata per la challenge tls-alpn01)
- porta esterna 80 TCP -> 192.168.1.10 porta 80 TCP (che è la porta di Caddy usata per la challenge http-01)

Fatto questo, restartato Caddy, dovrebbe riuscire a superare la challenge e a scaricare il certificato corretto:

https://tuodominio:9000 per accesso da esterno
http://192.168.1.10:8123 per accesso da LAN

**Reply by snakuzzo** [reply]
permettimi poi di dire che se è troppo complicato per te, lascia perdere e passa a Cloudflare

**Reply by basebuzz** [reply]
Ciao, allora prima di tutto nulla è complicato per me quindi se vuoi aiutarmi puoi farlo senza stare a pensare che non sia in grado di risolvere. ho fatto cose pià complesse di queste nella vita :) Secondo punto molto importante: funzionava tutto, quindi sono riuscito correttamente a configurarlo, non è una questione di capacità, è andato per un bel po e poi a smesso di funzionare. per questo sono qui nuovamente, per capire cosa blocca il funzionamento.
ho visto che hai messo un esempio del caddyfile, non credo di averlo simile, dove si trova esattamente questo file di configurazione? se raggiungo quello nelle impostazioni di Caddy e clicclo su yaml non è scritto in questo modo, sembra una conf di default.
Grazie per l'aiuto.

**Reply by basebuzz** [reply]
ok ora funziona tutto nuovamente.
l'unica cosa che ho fatto è stato impostare la regola sul router:
- porta esterna 443 TCP -> 192.168.1.10 porta 443 TCP (che è la porta di Caddy usata per la challenge tls-alpn01)
Se metto anche la regola per la porta 80 accedo ad Ha ma sia da rete interna che da esterna non mi trova più un sacco di dispositivi.
Quindi a posto così, vediamo alla prossima challenge cosa succederà.
Caddy non da più errori.
Grazie 1000 per la pazienza.

**Reply by snakuzzo** [reply]
Caddy ripasserà la challenge circa 20-30 gg prima della scadenza. Se non fai casini con le regole di port forwarding non dovresti più preoccuparti di nulla
